File: backend/trackline/models/games.py
from datetime import datetime
from enum import Enum
from typing import Dict, List

from pydantic import Field

from trackline.models.base import BaseModel, IdentifiableModel
from trackline.utils.fields import StringId


class GameState(str, Enum):
    WAITING_FOR_PLAYERS = "waiting_for_players"
    STARTED = "started"
    GUESSING = "guessing"
    SCORING = "scoring"
    COMPLETED = "completed"
    ABORTED = "aborted"


class Track(BaseModel):
    spotify_id: str
    title: str
    artists: str
    release_year: int
    image_url: str | None


class Player(BaseModel):
    user_id: StringId
    is_game_master: bool = False
    tokens: int = 0
    timeline: List[Track] = []


class Guess(BaseModel):
    creation_time: datetime = Field(default_factory=datetime.now)
    position: int | None = None
    release_year: int | None = None


class CategoryScoring(BaseModel):
    winner: str | None
    tokens_delta: Dict[str, int]


class TurnScoring(BaseModel):
    position: CategoryScoring
    release_year: CategoryScoring


class Turn(BaseModel):
    creation_time: datetime = Field(default_factory=datetime.now)
    active_user_id: StringId
    track: Track
    guesses: Dict[str, Guess] = {}
    scoring: TurnScoring | None
    completed_by: List[str] = []


class GameSettings(BaseModel):
    spotify_market: str
    playlist_ids: List[str]
    initial_tokens: int
    timeline_length: int
    guess_timeout: int


class Game(IdentifiableModel):
    creation_time: datetime = Field(default_factory=datetime.now)
    settings: GameSettings
    state: GameState = GameState.WAITING_FOR_PLAYERS
    turns: List[Turn] = []
    players: List[Player] = []
    discarded_track_ids: List[str] = []

    def get_player(self, user_id: str) -> Player | None:
        return next((p for p in self.players if p.user_id == user_id), None)

    def get_active_player(self) -> Player | None:
        if not self.turns:
            return None

        active_user_id = self.turns[-1].active_user_id
        return self.get_player(active_user_id)

File: frontend/src/types/games.ts
export enum GameState {
  WaitingForPlayers = "waiting_for_players",
  Started = "started",
  Guessing = "guessing",
  Scoring = "scoring",
  Completed = "completed",
  Aborted = "aborted",
}

export enum ArtistMatchMode {
  All = "all",
  One = "one",
}

export enum TitleMatchMode {
  Full = "full",
  Main = "main",
}

export enum CreditsStrictness {
  Exact = "exact",
  Strict = "strict",
  Moderate = "moderate",
  Relaxed = "relaxed",
}

export enum CorrectionProposalState {
  Voting = "VOTING",
  Accepted = "ACCEPTED",
  Rejected = "REJECTED",
}

export interface Track {
  spotifyId: string;
  title: string;
  artists: string[];
  releaseYear: number;
  imageUrl: string;
}

export interface Player {
  userId: string;
  isGameMaster: boolean;
  tokens: number;
  timeline: Track[];
}

export interface Guess {
  creationTime: string;
  userId: string;
  tokenCost: number;
}

export interface ReleaseYearGuess extends Guess {
  position: number;
  year: number;
}

export interface CreditsGuess extends Guess {
  artists: string[];
  title: string;
}

export interface TurnPass {
  userId: string;
  creationTime: string;
}

export interface CorrectionProposalVote {
  userId: string;
  agree: boolean;
  creationTime: string;
}

export interface CorrectionProposal {
  createdBy: string;
  creationTime: string;
  state: CorrectionProposalState;
  releaseYear: number;
  votes: CorrectionProposalVote[];
}

export interface TokenGain {
  refund: number;
  rewardTheoretical: number;
  rewardEffective: number;
}

interface Scoring {
  winner: string;
  correctGuesses: string[];
  tokenGains: { [userId: string]: TokenGain };
}

export interface ReleaseYearScoring {
  position: Scoring;
  year: Scoring;
}

export interface CreditsScoring extends Scoring {
  similarityScores: { [userId: string]: number };
}

export interface TurnScoring {
  releaseYear: ReleaseYearScoring;
  credits: CreditsScoring;
}

export interface Turn {
  revisionId: string;
  creationTime: string;
  activeUserId: string;
  track: Track;
  catchUpTokenGain: { [userId: string]: number };
  guesses: {
    releaseYear: ReleaseYearGuess[];
    credits: CreditsGuess[];
  };
  passes: TurnPass[];
  scoring: TurnScoring;
  correctionProposal: CorrectionProposal;
  completedBy: string[];
}

export interface GameSettings {
  artistsMatchMode: ArtistMatchMode;
  creditsSimilarityThreshold: number;
  guessTimeout: number;
  initialTokens: number;
  maxTokens: number;
  playlistIds: string[];
  spotifyMarket: string;
  timelineLength: number;
  titleMatchMode: TitleMatchMode;
}

export interface Game {
  id: string;
  creationTime: string;
  settings: GameSettings;
  state: GameState;
  turns: Turn[];
  players: Player[];
}
